### PR TITLE
Custom providers #104

### PR DESCRIPTION
**Pull request by jojomartius**
Enables custom providers see #104

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -321,6 +321,36 @@ use Cake\Event\EventManager;
...
+// Instantiate your own CollectionFactory with
+...
+public function middleware($middlewareQueue): \Cake\Http\MiddlewareQueue
+...
+$cf = new CollectionFactory();
+$cf->register(\App\Authenticator\CustomProvider::NAME, \App\Authenticator\CustomProvider::class);
+...
+
+// Add this to your configuration
+serviceConfig.collectionFactory => $cf,
```
Comment (ADmad) [inline README.md:343]: Don't nest `collectionFactory` under `serviceConfig`. Keep it as a separate top level config. Additional key inside the service options argument could cause problems in future.

Comment (jojomartius) [reply README.md]: yes, moved

File: src/Middleware/SocialAuthMiddleware.php [hunk 2]
```diff
@@ -173,9 +173,9 @@ class SocialAuthMiddleware implements MiddlewareInterface, EventDispatcherInterf
     /**
      * Constructor.
      *
-     * @param array $config Configuration.
-     * @param \Cake\Event\EventManager|null $eventManager Event manager instance.
-     * @param \SocialConnect\Provider\Session\SessionInterface|null $session Session handler for SocialConnect Service
+     * @param array                                                 $config       Configuration.
```
Comment (ADmad) [inline src/Middleware/SocialAuthMiddleware.php:176]: All these unnecessary whitespace changes to docblocks need to be reverted.

Comment (jojomartius) [reply src/Middleware/SocialAuthMiddleware.php]: your absolut right, no idea where this comes from. so i just run 
`composer run-script cs-fix`
before committing and corrected all whitespace changes that i saw, hope thats fine now. BTW, is there an official way how to format and adjust code style?

Comment (ADmad) [reply src/Middleware/SocialAuthMiddleware.php]: Not sure `composer run-script cs-fix` messed up the formatting instead of fixing it. I will look into it later.

There are still lot of unnecessary formatting changes remaining. Instead of trying to fix them just drop the commits you made after running CS fixer. There was only one CS error before that which you could fix manually.

**Comment by jojomartius** [comment]
Ok, now i tried to revert to the last state and anytime the editor tried to do some code formatting i removed it from commit.

**Comment by ADmad** [comment]
Instead of just dropping the extra commits which messed up the formatting you keep adding more commit :slightly_smiling_face:.

Use `git rebase -i HEAD~8`. It will open your configured editor listing the last 8 commits. In there simply remove all the commits listed after the one where you moved the config to top level and save/close the editor. That will drop all the crap.

Then use `git push -f origin aircargobook:master` to update the PR. After that fix any reported issues manually, commit and push again normally.

**Comment by jojomartius** [comment]
you are so right, should be done. PR updated

**Comment by ADmad** [comment]
Great, you can now fix the couple of issues reported (ignore the static analysis error).

File: README.md [hunk 3]
```diff
@@ -321,6 +321,35 @@ use Cake\Event\EventManager;
 EventManager::instance()->on(new SocialAuthListener());
 ```
 
+### Extend with custom providers
+
+In order to enable custom or not preconfigured provideres you can extend the middleware configuration with *collectionFactory* and create your own instance of *SocialConnect\Auth\CollectionFactory*.
```
Comment (ADmad) [inline README.md:326]:
```suggestion
In order to enable custom or not preconfigured providers you can extend the middleware configuration with `collectionFactory` and create your own instance of `SocialConnect\Auth\CollectionFactory`.
```

File: README.md [hunk 4]
```diff
@@ -321,6 +321,35 @@ use Cake\Event\EventManager;
...
+$cf->register(\App\Authenticator\CustomProvider::NAME, \App\Authenticator\CustomProvider::class);
+...
+
+// Add this to your configuration
+'collectionFactory' => $cf,
+
+
+// src/Authenticator\CustomProvider.php
+/*
+    You will find a lot of examples invendor/socialconnect/auth/src/OpenIDConnect/Provider/
```
Comment (ADmad) [inline README.md:348]:
```suggestion
    You will find a lot of examples in vendor/socialconnect/auth/src/OpenIDConnect/Provider/
```

File: README.md [hunk 5]
```diff
@@ -321,6 +321,35 @@ use Cake\Event\EventManager;
...
+*/
+```
+
+
```
Comment (ADmad) [inline README.md:352]:
```suggestion
```
